core/src/main/scala/io/projectglow/sql/util/ExpectsGenotypeFields.scala

```diff
@@ -21,36 +21,58 @@ import org.apache.spark.sql.catalyst.analysis.{TypeCheckResult, UnresolvedExcept
 import org.apache.spark.sql.catalyst.expressions.{Expression, Unevaluable}
 import org.apache.spark.sql.types.{ArrayType, DataType, StructField, StructType}
 
+case class GenotypeInfo(size: Int, requiredFieldIndices: Seq[Int], optionalFieldIndices: Seq[Int])
+
 /**
  * A trait to simplify type checking and reading for expressions that operate on arrays of genotype
  * data with the expectation that certain fields exists.
+ *
+ * Note: This trait introduces complexity during resolution and analysis, and prevents
+ * nested column pruning. Prefer writing new functions as rewrites when possible.
  */
-trait ExpectsGenotypeFields extends Expression {
+@deprecated trait ExpectsGenotypeFields extends Expression {
+  def genotypeInfo: Option[GenotypeInfo]
+  final def getGenotypeInfo: GenotypeInfo = {
```

Comment:
It's not strictly necessary, but I thought it was helpful to have a wrapper that forcefully gets the `GenotypeInfo` so that it's more clear that subclasses don't need to guard against the case where it's `None` during execution.